### PR TITLE
More features added to enable pipelining of messages in updates

### PR DESCRIPTION
**Pull request by besuikerd**
I Was experimenting around with your module and came across a few use cases where I needed a few more functions regarding updates. These are the functions I implemented:

* `andThen`
* `filter`
* `addCmd`
* `batch`

I added a [test example](https://github.com/besuikerd/elm-update-extra/blob/master/test/Counter/Main.elm) to test out these new operators.

I tried to match the method names to match commonly used names in other modules. `andThen` does the same thing as `pipeUpdate`, but with its arguments flipped. This allows it to be used with the 'pipeline operator' `|>`.

example usage:
```elm
update msg model =
  model ! []
    |> andThen update SomeMessage
    |> andThen update SomeOtherMessage
    ...
```

`filter` is used to only send the messages if the predicate evaluates to true

example usage:
```elm
update msg model =
  case msg of
    SomeMessage i ->
      model ! []
        |> filter (i > 10)
          ( \state -> state
              |> andThen update BiggerThanTen
              |> andThen update AnotherMessage
              |> andThen update EvenMoreMessages
          )
        |> andThen update AlwaysTriggeredAfterPredicate

```

`batch` is the operator as discussed by #2, #4. I think `batch` is a good fit considering `Cmd` and `Sub` also use them to concatenate. These are also used with `|>`:

```elm
update msg model = model ! []
  |> batch update
    [ AMessage
    , AnotherMessage
    , AThirdMessage
    ]
```

`addCmd` is added as requested by #3.

EDIT: updated `andThen` to match the signatures

### CODE REVIEW

**Comment by ccapndave** [comment]
Thanks very much, this is great! I'm off on holiday at the moment but I'll have a proper look at this when I get back and get it all merged and published.

Thanks again!

File: src/Update/Extra.elm [hunk 1]
```diff
@@ -25,3 +35,87 @@ pipeUpdate (model, cmd) f =
...
+update msg model =
+  model ! []
+    |> andThen update SomeMessage
+    |> andThen update SomeOtherMessage
+    ...
+```
+-}
+andThen : (msg -> model -> (model, Cmd msg)) -> msg -> (model, Cmd msg) -> (model, Cmd msg)
+andThen update msg (model, cmd) =
+  let
```
Comment (jvoigtlaender) [inline src/Update/Extra.elm]: andThen update msg = flip pipeUpdate (update msg)

**Comment by jvoigtlaender** [comment]
One subtlety about the choice of name `batch`: In `Cmd.batch` and `Sub.batch` the "list" is actually used as a "bag", that is, without giving any relevance to the order of items batched. Indeed, in Elm `Cmd.batch` and `Sub.batch` are to be considered semantically equivalent to `List.reverse >> Cmd.batch` and `List.reverse >> Sub.batch`, respectively (in the case of `Cmd`, due to the asynchronicity of `Cmd`s). But the `batch` function here does not share this property. Due to its immediate effect on the model the order of items is suddenly relevant. That could become the origin of confusion in the long run.

**Comment by besuikerd** [comment]
@jvoigtlaender Good catch, the sequencing should be visible in the name.

Alternatives could be `sequence` or `foldl`. I am fine with either of these

**Comment by jvoigtlaender** [comment]
Between these two, I'm very much in favour of "sequence".

**Comment by ccapndave** [comment]
I agree - sequence is clearer

File: src/Update/Extra.elm [hunk 2]
```diff
@@ -101,16 +101,16 @@ addCmd cmd' (model, cmd) = (model, Cmd.batch [cmd, cmd'])
 {-| allows you to attach multiple messages to an update at once.
 
 ```elm
-update msg model = model ! []
+sequence msg model = model ! []
   |> batch update
     [ AMessage
```
Comment (jvoigtlaender) [inline src/Update/Extra.elm]: Looks like you renamed the wrong thing here.

**Comment by ccapndave** [comment]
This looks great.  Now that you have written the `andThen` function, I can actually see that this would be a more Elmy name than `pipeUpdate`.  Therefore I propose that we simply get rid of `pipeUpdate` and redefine `:>` in terms of `andThen` as I'm 99% sure that everyone is using the infix version anyway (plus `elm-package` will make sure that nothing breaks).  Also I definitely think that `batch` should be renamed `sequence` (in the tests too).

**Comment by besuikerd** [comment]
Cool, I will change that aswell :ok_hand:

**Comment by besuikerd** [comment]
Hmm `:>` cannot be expressed in terms of `andThen`. The signature of `andThen` expects the whole update function and message separately, that information is lost when the infix operator is used since the message is applied to the update function first there. 

So there needs to be an intermediary implementation, the one of `pipeUpdate`. Shall we keep the `pipeUpdate` implementation or hide it?

**Comment by jvoigtlaender** [comment]
I think you are wrong here. The two functions should be interdefinable. Just use an "empty message" of type ().

**Comment by ccapndave** [comment]
I think `pipeUpdate = flip andThen` doesn't it?

**Comment by jvoigtlaender** [comment]
Well, that isn't type-correct.

**Comment by besuikerd** [comment]
They are but the other way around, `andThen` can be expressed in terms of `:>` operator, but it feels wrong to write an implementation on an operator.

`andThen` no longer is simply the flipped version of `pipeUpdate`. the signature is now

```elm
(msg -> model -> (model, Cmd msg)) -> msg -> (model, Cmd msg) -> (model, Cmd msg)
```

**Comment by jvoigtlaender** [comment]
When I say interdefinable, I mean each can be expressed in terms of the other. :smile:

**Comment by besuikerd** [comment]
Haha ofcourse, misread your comment :innocent:

**Comment by jvoigtlaender** [comment]
```elm
pipeUpdate (model, cmd) f = andThen (\() -> f) () (model, cmd)
```

**Comment by ccapndave** [comment]
Just realised I've been looking at the wrong commit in this PR the whole time ;)

**Comment by jvoigtlaender** [comment]
Oh, even in the latest commit, the documentation of `andThen` still says:
> Identical to `pipeUpdate`, but with reversed arguments to be able to use it with the pipeline operator (`|>`)

**Comment by jvoigtlaender** [comment]
Maybe the `andThen` shouldn't exist. When is it more useful than the `(|>)` operator?

**Comment by besuikerd** [comment]
Mainly because you don't have to remember another operator. While operators can reduce the size of your code, they make your code less readable

**Comment by ccapndave** [comment]
Is the only reason for separating out the `msg` in the signature so that you don't need brackets after `andThen`?

**Comment by besuikerd** [comment]
@ccapndave correct

**Comment by ccapndave** [comment]
Ok - well, I'm easy.  Personally I don't mind the brackets after `andThen`, but really whether or not to keep `pipeUpdate` is an implementation detail as people will either be using `:>` or `|> andThen`.

**Comment by jvoigtlaender** [comment]
A separate concern: the type variables should be used consistently in the signatures, because they will show up in the documentation pages, and it's odd to read there
```elm
(:>) : (m, Cmd a) -> (m -> (m, Cmd a)) -> (m, Cmd a)
```
but
```elm
andThen : (msg -> model -> (model, Cmd msg)) -> msg -> (model, Cmd msg) -> (model, Cmd msg)
```

**Comment by besuikerd** [comment]
I will fix the same type variable names aswell.

Had not event thought of feeding it a unit message, nice! For it to fit, the type signature of `andThen` need to be relaxed a bit more:

```elm
andThen : (msg -> model -> (model, Cmd msg')) -> msg -> (model, Cmd msg') -> (model, Cmd msg')
```

**Comment by jvoigtlaender** [comment]
Right, about the need for `msg'` then. But then, is it worth that? Is it a value in itself to define `(:>)` via `andThen`? Maybe it's better to keep the simpler signature of `andThen`, and simply define `(:>)` independently.

**Comment by ccapndave** [comment]
Agreed; given how the typing has panned out we might as well define `:>` separately in the `Infix` module.

**Comment by ccapndave** [comment]
Good work!

**Comment by ccapndave** [comment]
http://package.elm-lang.org/packages/ccapndave/elm-update-extra/2.0.0

**Comment by ccapndave** [comment]
Kudos on this `filter` function, by the way.  I didn't know how much I needed it until it was there 👍

**Comment by besuikerd** [comment]
haha, cheers :beers:

**Pull request merged**